src/code/z_message_PAL.c

```diff
@@ -1,66 +1,3306 @@
+// original name: onpu_buff
+u8 sOcarinaNoteBuf[12] = { 0 };
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010BD58.s")
+s16 sOcarinaNotesAlphaValues[9] = { 0 };
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010BD88.s")
+// Maps the ocarina song order to the quest item order
+s16 gOcarinaSongItemMap[] = {
+    OCARINA_SONG_MINUET,   OCARINA_SONG_BOLERO,  OCARINA_SONG_SERENADE, OCARINA_SONG_REQUIEM,
+    OCARINA_SONG_NOCTURNE, OCARINA_SONG_PRELUDE, OCARINA_SONG_LULLABY,  OCARINA_SONG_EPONAS,
+    OCARINA_SONG_SARIAS,   OCARINA_SONG_SUNS,    OCARINA_SONG_TIME,     OCARINA_SONG_STORMS,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010BDBC.s")
+s32 sCharTexSize;
+s32 sCharTexScale;
+s16 sOcarinaNoteAPrimR;
+s16 sOcarinaNoteAPrimB;
+s16 sOcarinaNoteAPrimG;
+s16 sOcarinaNoteAEnvR;
+s16 sOcarinaNoteAEnvB;
+s16 sOcarinaNoteAEnvG;
+s16 sOcarinaNoteCPrimR;
+s16 sOcarinaNoteCPrimB;
+s16 sOcarinaNoteCPrimG;
+s16 sOcarinaNoteCEnvR;
+s16 sOcarinaNoteCEnvB;
+s16 sOcarinaNoteCEnvG;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010BED8.s")
+void Message_ResetOcarinaNoteState(void) {
+    R_OCARINA_NOTES_YPOS(0) = 189;
+    R_OCARINA_NOTES_YPOS(1) = 184;
+    R_OCARINA_NOTES_YPOS(2) = 179;
+    R_OCARINA_NOTES_YPOS(3) = 174;
+    R_OCARINA_NOTES_YPOS(4) = 169;
+    sOcarinaNoteBuf[0] = 0xFF;
+    sOcarinaNotesAlphaValues[0] = sOcarinaNotesAlphaValues[1] = sOcarinaNotesAlphaValues[2] =
+        sOcarinaNotesAlphaValues[3] = sOcarinaNotesAlphaValues[4] = sOcarinaNotesAlphaValues[5] =
+            sOcarinaNotesAlphaValues[6] = sOcarinaNotesAlphaValues[7] = sOcarinaNotesAlphaValues[8] = 0;
+    sOcarinaNoteAPrimR = 80;
+    sOcarinaNoteAPrimG = 255;
+    sOcarinaNoteAPrimB = 150;
+    sOcarinaNoteAEnvR = 10;
+    sOcarinaNoteAEnvG = 10;
+    sOcarinaNoteAEnvB = 10;
+    sOcarinaNoteCPrimR = 255;
+    sOcarinaNoteCPrimG = 255;
+    sOcarinaNoteCPrimB = 50;
+    sOcarinaNoteCEnvR = 10;
+    sOcarinaNoteCEnvG = 10;
+    sOcarinaNoteCEnvB = 10;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010C358.s")
+void Message_UpdateOcarinaGame(GlobalContext* globalCtx) {
+    MessageContext* msgCtx = &globalCtx->msgCtx;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010C39C.s")
+    globalCtx->msgCtx.msgMode++;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message_PAL/func_8010F2CC.s")
+    if (globalCtx->msgCtx.msgMode == MSGMODE_MEMORY_GAME_PLAYER_PLAYING) { // Player playing
+        Audio_OcaSetInstrument(1);
+        msgCtx->ocarinaStaff = Audio_OcaGetPlayingStaff();
+        msgCtx->ocarinaStaff->pos = sOcarinaNoteBufPos = 0;
+        func_800ECC04((1 << OCARINA_SONG_MEMORY_GAME) + 0x8000);
+        msgCtx->textDrawPos = msgCtx->decodedTextLen;
+    } else if (msgCtx->msgMode == MSGMODE_MEMORY_GAME_RIGHT_SKULLKID_PLAYING) { // Skull kids playing
+        Audio_OcaSetInstrument(6);
+        msgCtx->ocarinaStaff = Audio_OcaGetDisplayingStaff();
```

Comment:
probably not in scope of this PR, but I wouldnt mind just saying Ocarina for these instead of Oca